core/executor/src/wasmtime/trampoline.rs

```diff
@@ -0,0 +1,312 @@
+// Copyright 2019 Parity Technologies (UK) Ltd.
+// This file is part of Substrate.
+
+// Substrate is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+
+// Substrate is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with Substrate.  If not, see <http://www.gnu.org/licenses/>.
+
+//! The trampoline is the dynamically generated entry point to a runtime host call.
+//!
+//! This code is based on and large parts are copied from wasmtime's
+//! wasmtime-api/src/trampoline/func.rs.
+
+use cranelift_codegen::{Context, binemit, ir, isa};
+use cranelift_codegen::ir::{InstBuilder, StackSlotData, StackSlotKind, TrapCode};
+use cranelift_frontend::{FunctionBuilder, FunctionBuilderContext};
+use wasmtime_jit::{CodeMemory, Compiler};
+use wasmtime_runtime::{VMContext, VMFunctionBody};
+use wasm_interface::{HostFunctions, Function, Value, ValueType};
+use std::{cmp, ptr};
+
+use crate::error::{Error, Result};
+use crate::wasmtime::function_executor::{FunctionExecutorState, FunctionExecutor};
+
+/// The top-level host state of the "env" module. This state is used by the trampoline function to
+/// construct a `FunctionExecutor` which can execute the host call.
+pub struct EnvState {
+	externals: &'static [&'static dyn Function],
+	compiler: Compiler,
+	// The code memory must be kept around on the state to prevent it from being dropped.
+	#[allow(dead_code)]
+	code_memory: CodeMemory,
+	trap: Option<Error>,
+	/// The executor state stored across host calls during a single Wasm runtime call.
+	/// During a runtime call, this MUST be `Some`.
+	pub executor_state: Option<FunctionExecutorState>,
+}
+
+impl EnvState {
+	/// Construct a new `EnvState` which owns the given code memory.
+	pub fn new<HF: HostFunctions>(code_memory: CodeMemory, compiler: Compiler) -> Self {
+		EnvState {
+			externals: HF::functions(),
+			trap: None,
+			compiler,
+			code_memory,
+			executor_state: None,
+		}
+	}
+
+	/// Resets the trap error to None and returns the current value.
+	pub fn take_trap(&mut self) -> Option<Error> {
+		self.trap.take()
+	}
+}
+
+/// This is called by the dynamically generated trampoline taking the function index and reference
+/// to the call arguments on the stack as arguments. Returns 0 on success and 1 on an error.
+unsafe extern "C" fn stub_fn(vmctx: *mut VMContext, func_index: u32, values_vec: *mut i64) -> u32 {
+	if let Some(state) = (*vmctx).host_state().downcast_mut::<EnvState>() {
+		match stub_fn_inner(
```

Comment:
Wasmtime seems to be catching this just fine since it sets up a signal handler, so I don't see the need for an explicit `catch_unwind` except that we might be able to give a better error message. (The default is `Err(Other("Wasm execution trapped: wasm trap: unreachable, source location: @1395"))`). And we would only be able to give a better error message if it panics in the external function, not the stub code. Thoughts?

Reply:
Basti is right here since theoretically unwinding into FFI is UB. There was an attempt to fix that (well it would be still UB, not 100% sure, IANAL) in 1.24, but it got reverted in 1.24.1.

Because it is a bit sketchy I'd avoid this.

Reply:
My theory is that in most cases UB would be manifested in just an `unreachable`, which would be translated in `ud2` and thus be caught by the signal handler (Happy to be proven wrong though.). 

I'd be strongly encourage to not depend on any behavior of UB.